src/rust/catloop/mod.rs

```diff
@@ -0,0 +1,521 @@
+    pub fn sgaalloc(&self, size: usize) -> Result<demi_sgarray_t, Fail> {
+        self.catmem.borrow_mut().alloc_sgarray(size)
+    }
+
+    /// Releases a scatter-gather array.
+    pub fn sgafree(&self, mut sga: demi_sgarray_t) -> Result<(), Fail> {
+        // Sanity check length of scatter-gather array.
+        // ToDo: Drop this check once we support scatter-gather arrays with multiple segments.
+        if sga.sga_numsegs != 1 {
+            return Err(Fail::new(libc::EINVAL, "demi_sgarray_t has invalid segment count"));
+        }
+
+        // Check for magic close message.
+        if sga.sga_segs[0].sgaseg_len == 0 {
+            sga.sga_segs[0].sgaseg_len = mem::size_of_val(&Self::MAGIC_CLOSE) as u32;
+            // Sanity check.
+            assert_eq!(self.is_magic_close(&mut sga), true);
+        }
+
+        self.catmem.borrow_mut().free_sgarray(sga)
+    }
+
+    pub fn schedule(&mut self, qt: QToken) -> Result<SchedulerHandle, Fail> {
+        match self.qts.get(&qt) {
+            Some((demi_opcode_t::DEMI_OPC_ACCEPT, _)) | Some((demi_opcode_t::DEMI_OPC_CONNECT, _)) => {
+                match self.scheduler.from_raw_handle(qt.into()) {
+                    Some(handle) => Ok(handle),
+                    None => return Err(Fail::new(libc::EINVAL, "invalid queue token")),
+                }
+            },
+            Some((demi_opcode_t::DEMI_OPC_PUSH, _)) | Some((demi_opcode_t::DEMI_OPC_POP, _)) => {
+                self.catmem.borrow_mut().schedule(qt)
+            },
+            _ => return Err(Fail::new(libc::EINVAL, "invalid queue token")),
+        }
+    }
+
+    pub fn pack_result(&mut self, handle: SchedulerHandle, qt: QToken) -> Result<demi_qresult_t, Fail> {
+        match self.qts.remove(&qt) {
+            Some((demi_opcode_t::DEMI_OPC_ACCEPT, _)) | Some((demi_opcode_t::DEMI_OPC_CONNECT, _)) => {
+                let (qd, r): (QDesc, OperationResult) = self.take_result(handle);
+
+                match r {
+                    OperationResult::Connect((remote, ref duplex_pipe)) => {
+                        _ = self.sockets.insert(qd, Some((remote, duplex_pipe.clone())))
+                    },
+                    OperationResult::Accept((new_qd, (remote, ref duplex_pipe))) => {
+                        self.sockets.insert(new_qd, Some((remote, duplex_pipe.clone())));
+                    },
+                    _ => {},
+                };
+
+                return Ok(pack_result(r, qd, qt.into()));
+            },
+            Some((demi_opcode_t::DEMI_OPC_PUSH, qd)) | Some((demi_opcode_t::DEMI_OPC_POP, qd)) => {
+                let mut qr: demi_qresult_t = self.catmem.borrow_mut().pack_result(handle, qt)?;
+                qr.qr_qd = qd.into();
+
+                if qr.qr_opcode == demi_opcode_t::DEMI_OPC_POP {
+                    let mut sga: &mut demi_sgarray_t = unsafe { &mut qr.qr_value.sga };
+
+                    // Check for close.
+                    if self.is_magic_close(sga) {
```

Comment:
What if someone legitimately sends the MAGIC_CLOSE bytes?  Is that byte sequence not allowed?

Reply:
It is allowed and will not yield to an erroneous close. The rationale is that a close sequence consists of:

- A scatter-gather array with zero-length. Note that a user cannot allocate this by construction, from the `sga_alloc()` function.
- A payload with the `MAGIC_CLOSE` sequence.

Reply:
How does that first point work?  In this code here, `is_magic_close` checks for the magic payload (with length correct for normal data) and adjusts the length to the bogus "magic close" value if the payload is the magic one.  It looks like this code gets called for things other than close, so if someone happens to send the magic payload, won't this turn it into a close?

Reply:
# What is Implemented

- Legit scatter-Gather arrays are allocated from Catmem LibOS, and have 1 segment:

https://github.com/demikernel/demikernel/blob/5596a9173071634833a6c0f4720f3d70f24fd0d0/src/rust/catmem/mod.rs#L163-L189

- A `close()` operation in Catloop LibOS translates to a `push()` request in Catmem LibOS (note that this permalink may change because it refers to the branch in the PR):

https://github.com/demikernel/demikernel/blob/c3970cc85e5c40f0af39e531051f7fbf2b57e67a/src/rust/catloop/mod.rs#L263-L297

- The magic payload is cooked there:

https://github.com/demikernel/demikernel/blob/c3970cc85e5c40f0af39e531051f7fbf2b57e67a/src/rust/catloop/mod.rs#L454-L467

- When the the message containing the magic payload arrives at the other end, we truncate the length of the buffer to zero when creating the `demi_qresult_t`:

https://github.com/demikernel/demikernel/blob/c3970cc85e5c40f0af39e531051f7fbf2b57e67a/src/rust/catloop/mod.rs#L369-L409

- At the receiver side, double free is prevented:

https://github.com/demikernel/demikernel/blob/c3970cc85e5c40f0af39e531051f7fbf2b57e67a/src/rust/catloop/mod.rs#L336-L352

Reply:
Yes, that's exactly what I thought was happening.  My question is: what is different about the message that gets sent by `close()` (which calls `cook_magic_close()` to generate a scatter-gather array with the `MAGIC_CLOSE` message, and then pushes that message) from a normal user-generated message that just happens to contain the same sequence of bytes as the `MAGIC_CLOSE` message does?

Or to put it another way, how does the code at "the other end" detect that the `close()` generated message isn't a user-generated message that just happens to have those magic bytes?  The code in pack_result I flagged originally appears to just be looking for that byte sequence and setting the length to zero if it sees it.  That's my question.

Reply:
@BrianZill 

Oh, yes. I understand. Thanks for spotting this. I was focused on the application not crashing due to a segmentation faut on a zero-length buffer.

Currently, in the case where the other end gets a `MAGIC_CLOSE` sequence, it has no way to distinguish a `close()` from an usual `push()` . The alternatives that came up for solving this are:

- The end process that received this `MAGIC_CLOSE` sequence, waits for some timeout to decide whether or not that was a legit close. This does not fix the problem, because we are making assumptions on the "pace of the communication".
- We encode the `MAGIC_CLOSE` sequence as multiple messages, each of each with a different payload. This reduces the chance of an erroneous `close()` but does not fix the problem.
- We implement a four-way handshake like we have in TCP, to reduce even further the chance of having an erroneous `close()`. This also does not fix the problem, because it is a legit communication to have a sequence of `MAGIC_CLOSE`s being passed around. In TCP it works because each payload has an additional header message, where the `FIN` bit is set.

Let me know if you have any additional ideas that would permanently fix the problem and not incur in additional overheads.

Reply:
Well... (you're going to love this one)...

We could re-architect the system to pass buffers instead of bytes through the `SharedRingBuffer`s.  I.e., Make them `SharedRingBuffer<usize>` rather than u8, change catmem to pass a reference to the `DemiBuffer` instead of looping through the individual bytes and pushing each through the queue individually, and change catloop to allocate all of its `DemiBuffer`s in the shared memory space so the other end can dereference the pointers and get something meaningful.  If the shared memory space could also be used to map the DPDK-backed `DemiBuffer`s, then data received off the network could be passed through catloop to another process with zero-copy.  That would be super-fast.

Oh, and this would make an out-of-band signal for "close" trivial, as we could just use all zeroes instead of a valid `DemiBuffer` address to indicate the channel was closing.

I suppose a less invasive change would be to make the `SharedRingBuffer`s pass a u16, but normally just use half of that to pass a u8 as you do now.  But that would give you lots of extra bits to pass out-of-band information, such as the close indication.

Or, since you are currently passing each u8 individually, you could add your own special multi-byte encoding.  Pick one value to be your special "close" byte (or starting byte of a multi-byte "close" sequence for atomicity reasons) and double it (at the sender, remove extra byte at the receiver) in the normal case but actually send the single byte to "close".  That would be more complicated than passing u16s though but would use up less shared memory space.

Reply:
> Well... (you're going to love this one)...

Oh, definitely :)

I have changed Catmem LibOS to use a `SharedRingBuffer<u16>` and use the low byte for data and the high one for control information. Normal pushes, always fill in the high byte with zeros, whereas a "close" push fills in the high byte. With this change all the `MAGIC_CLOSE` logic was removed from Catloop LibOS and now there is no ambiguity on "close()",

Reply:
Nice!